src/dataflow/src/source/mod.rs

```diff
@@ -1126,12 +1098,16 @@ where
             // Downgrade capability (if possible) before exiting
             consistency_info.downgrade_capability(&id, cap, source_info, &timestamp_histories);
 
-            // Ensure that we activate the source frequently enough to keep downgrading
-            // capabilities, even when no data has arrived
-            activator.activate_after(Duration::from_millis(
-                consistency_info.downgrade_capability_frequency,
-            ));
-            SourceStatus::Alive
+            // Schedule our next activation
+            match source_state.1 {
+                MessageProcessing::Yielded => activator.activate(),
+                MessageProcessing::YieldedWithDelay => activator.activate_after(
+                    Duration::from_millis(consistency_info.downgrade_capability_frequency),
+                ),
+                _ => (),
+            }
+
+            source_state.0
```

Comment:
Destructuring tuples is always less confusing IMO than indexing into them. Take it or leave it.

```suggestion
            let (source_status, processing_status) = source_state;
            // Schedule our next activation
            match processing_status {
                MessageProcessing::Yielded => activator.activate(),
                MessageProcessing::YieldedWithDelay => activator.activate_after(
                    Duration::from_millis(consistency_info.downgrade_capability_frequency),
                ),
                _ => (),
            }

            source_status
```

Reply:
Ah thank you, that does read better. 🙂 

I was hoping I could do:
```
(source_state, processing_state) = match ...
```
On line 986, but apparently Rust doesn't support that yet - only destructuring with a let.